src/hotspot/os/windows/memMapPrinter_windows.cpp

```diff
@@ -0,0 +1,256 @@
+    INDENT_BY(38);
+    st->print("%12zu", mInfo.RegionSize);
+    INDENT_BY(51);
+    st->print("%s", info.protectBuffer);
+    INDENT_BY(57);
+    st->print("%s-%s", info.stateBuffer, info.typeBuffer);
+    INDENT_BY(60);
+    st->print("%#9llx", reinterpret_cast<const unsigned long long>(mInfo.BaseAddress) - reinterpret_cast<const unsigned long long>(mInfo.AllocationBase));
+    INDENT_BY(72);
+    if (_session.print_nmt_info_for_region(mInfo.BaseAddress, static_cast<const char*>(mInfo.BaseAddress) + mInfo.RegionSize)) {
+      st->print(" ");
+    }
+    st->print_raw(info.fileName);
+  #undef INDENT_BY
+    st->cr();
+  }
+
+  void print_legend() const {
+    outputStream* st = _session.out();
+    st->print_cr("from, to, vsize: address range and size");
+    st->print_cr("prot:    protection:");
+    st->print_cr("             rwx: read / write / execute");
+    st->print_cr("             c: copy on write");
+    st->print_cr("             G: guard");
+    st->print_cr("             C: no cache");
+    st->print_cr("             W: write combine");
+    st->print_cr("             i: targets invalid");
+    st->print_cr("             n: targets noupdate");
+    st->print_cr("state:   region state and type:");
+    st->print_cr("             state: committed / reserved");
+    st->print_cr("             type: image / mapped / private");
+    st->print_cr("file:    file mapped, if mapping is not anonymous");
+    st->print_cr("vm info: VM information (requires NMT)");
+    {
+      streamIndentor si(st, 16);
+      _session.print_nmt_flag_legend();
+    }
+  }
+
+  void print_header() const {
+    outputStream* st = _session.out();
+    //            0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7
+    //            012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789
+    //            0x00007ffd5bd64000-0x00007ffd5c250000      5160960 r---  c-img 0x10d4000 C:\work\jdk\build\work-fastdebug\jdk\bin\server\jvm.dll
+    st->print_cr("from               to                        vsize prot  state    offset vm info/file");
+    st->print_cr("=============================================================================================================================");
+  }
+};
+
+void MemMapPrinter::pd_print_all_mappings(const MappingPrintSession& session) {
+
+    HANDLE hProcess = GetCurrentProcess();
+
+    MappingInfoPrinter printer(session);
+    MappingInfoSummary summary;
+
+    outputStream* const st = session.out();
+
+    printer.print_legend();
+    st->cr();
+    printer.print_header();
+
+    MEMORY_BASIC_INFORMATION mInfo;
+    MappingInfo info;
+
+    for (char* ptr = 0; VirtualQueryEx(hProcess, ptr, &mInfo, sizeof(mInfo)) == sizeof(mInfo); ptr += mInfo.RegionSize) {
```

Comment:
I would add a failsafe breakout in case we made a thinking error somewhere or the address space is insanely fragmented. Maybe break out after a million regions? Or, after a certain time (e.g. 10 seconds). I also would assert for RegionSize > 0

Reply:
done and done.